### PR TITLE
Payment backend

### PR DESCRIPTION
**Pull request by MangeshThakre**

### CODE REVIEW

File: backend/controller/paymentController.js [hunk 1]
```diff
@@ -109,12 +111,81 @@ const verifySubscription = asyncHandler(async (req, res, next) => {
...
+      currency: "INR",
+      description: description,
+    },
+  });
+
+  if (planResponse.error) {
+    return next(new CustomError(planResponse.error.description, 400));
+  }
+
+  const planeInfo = SubscriptionPlanModel({
```
Comment (shivamvijaywargi) [inline backend/controller/paymentController.js]: planInfo and not planeInfo

File: backend/middleware/asyncHandler.js [hunk 2]
```diff
@@ -1,3 +1,4 @@
+// this function will help to sent the error of  async task to error handler
```
Comment (shivamvijaywargi) [inline backend/middleware/asyncHandler.js]: Remove this comment

File: backend/middleware/authAdmin.js [hunk 3]
```diff
@@ -1,5 +1,8 @@
 const CustomError = require("../utils/customError");
 
+// this middleware will check the user is admin or not ,if not return error message.
+// Add this middleware in routes which you want only admin can access
+
 function authAdmin(req, res, next) {
   const role = req.user.role;
   if (role !== "ADMIN") {
```
Comment (shivamvijaywargi) [inline backend/middleware/authAdmin.js]: rewrite this to accept a array which we can check in the middleware

File: backend/router/paymentRouter.js [hunk 4]
```diff
@@ -3,13 +3,32 @@ const {
   createSubscription,
   getRazorpayApiKey,
   verifySubscription,
+  createPlan,
+  deletePlan,
+  updatePlan,
+  getPlans,
 } = require("../controller/paymentController");
 const jwtAuth = require("../middleware/jwtAuth.js");
+const authAdmin = require("../middleware/authAdmin");
```
Comment (shivamvijaywargi) [inline backend/router/paymentRouter.js]: After rewriting the auth middleware fix this as well

File: backend/router/paymentRouter.js [hunk 5]
```diff
@@ -3,13 +3,32 @@ const {
...
 
 paymentRouter.route("/subscribe").post(jwtAuth, createSubscription);
-paymentRouter.route("/razorpaykey").get(jwtAuth, getRazorpayApiKey);
 
-paymentRouter.route("/verify-subscription").post(jwtAuth, verifySubscription);
+paymentRouter.route("/rasorpaykey").get(jwtAuth, getRazorpayApiKey);
+
+paymentRouter
+  .route("/verify-subscription")
+  .post(jwtAuth, authAdmin, verifySubscription);
```
Comment (shivamvijaywargi) [inline backend/router/paymentRouter.js]: The auth middleware

File: backend/router/paymentRouter.js [hunk 6]
```diff
@@ -3,13 +3,32 @@ const {
...
+
+paymentRouter
+  .route("/plan")
+  .post(jwtAuth, authAdmin, createPlan)
```
Comment (shivamvijaywargi) [inline backend/router/paymentRouter.js]: The auth middleware

Comment (shivamvijaywargi) [reply backend/router/paymentRouter.js]: Check all the auth middlewares in the routes

File: backend/router/userRouter.js [hunk 7]
```diff
@@ -5,18 +5,28 @@ const checkUserSubscription = require("../middleware/ckeckUserSubscribtion.js");
 const {
```
Comment (shivamvijaywargi) [inline backend/router/userRouter.js:5]: Update the auth middleware here as well in this file

**Review by shivamvijaywargi — COMMENTED** [review]
Please check the reviews, fix them and merge

File: backend/controller/paymentController.js (hunk 1 above)
Comment (MangeshThakre) [reply backend/controller/paymentController.js]: done sir:)

File: backend/middleware/asyncHandler.js (hunk 2 above)
Comment (MangeshThakre) [reply backend/middleware/asyncHandler.js]: done sir please check

File: backend/middleware/authAdmin.js (hunk 3 above)
Comment (MangeshThakre) [reply backend/middleware/authAdmin.js]: done take a look

File: backend/router/paymentRouter.js (hunk 4 above)
Comment (MangeshThakre) [reply backend/router/paymentRouter.js]: replace with authorizeRoles

File: backend/router/paymentRouter.js (hunk 5 above)
Comment (MangeshThakre) [reply backend/router/paymentRouter.js]: done

File: backend/router/paymentRouter.js (hunk 6 above)
Comment (MangeshThakre) [reply backend/router/paymentRouter.js]: done

File: backend/router/userRouter.js (hunk 7 above)
Comment (MangeshThakre) [reply backend/router/userRouter.js:5]: done

**Review by shivamvijaywargi — CHANGES_REQUESTED** [review]
Please test the API endpoints before merging

**Review by shivamvijaywargi — APPROVED** [review]
Just test the endpoints before merging

**Pull request merged**